source/client/stream_decoder.h

```diff
@@ -64,10 +75,41 @@ class StreamDecoder : public Envoy::Http::StreamDecoder,
   void onPoolReady(Envoy::Http::StreamEncoder& encoder,
                    Envoy::Upstream::HostDescriptionConstSharedPtr host) override;
 
+  // The stream decoder isn't owned by anyone, and self destructs.
+  // Our pool implementation relies on a unique_ptr with a custom deleter.
+  // When we must play nice with a pool, we take ownership of that pointer through
+  // this call. The pointer will be released upon stream completion, resulting in
+  // either a pool recycle or a self destruct.
+  void takeOwnership(Nighthawk::PoolImpl<PoolableStreamDecoder>::PoolablePtr self) {
+    self_ = std::move(self);
+  }
+
+  void setCallerCompletionCallback(std::function<void()> caller_completion_callback) {
+    caller_completion_callback_ = std::move(caller_completion_callback);
+  }
+
+  void reset() {
+    complete_ = false;
+    have_first_body_byte_ = false;
+    milestone_tracker_.reset();
+    // TODO(oschaaf): this is kind of a hack to mark this here.
+    milestone_tracker_.markMilestone(milestone_request_start_);
+  }
+
+  void setMeasureLatencies(const bool measure_latencies) { measure_latencies_ = measure_latencies; }
+  void freeSelf() {
```

Comment:
This feels scary to me. I don't think an object should generally care who owns it; it should behave uniformly regardless. If you want some specific ownership/lifetime semantics for an object, you wrap it in something like a `unique_ptr` or `shared_ptr` that captures those semantics.

The `delete this` below is especially scary. It may be technically correct right now (see https://isocpp.org/wiki/faq/freestore-mgmt#delete-this), but I've always been told it should be avoided (https://doc.castsoftware.com/pages/viewpage.action?pageId=211748357).